### ISSUE TITLE
Memory usage starts to increase after external screen is connected

### ISSUE THREAD

**Issue by tarko** [awesomeWM/awesome#2983]
Following up the initial discussion at https://github.com/awesomeWM/awesome/issues/1382#issuecomment-576680991

I've noticed a memory usage issue with awesome - the process RSS just keeps growing into gigabytes and performance starts to degrade as well. It does not seem to be traditional memory leak but more like LUA issue as `collectgarbage("count")` keeps increasing and increasing while the amount of windows is pretty constant (~100 according to xwininfo). Obviously the window count goes up and down during normal usage but it's not constantly increasing.

I did some quick and dirty logging and plotted the data:

![awesome](https://user-images.githubusercontent.com/1047849/72807653-8e4a9180-3c60-11ea-97e6-9913deb43ae1.png)

- You can see awesome process RSS (that keeps increasing) and reported `collectgarbage("count")`
- The timeperiods where external screen is connected are highlighted with blue markers
- Skips in data are laptop sleep periods
- I also restarted awesome once (with all windows remaining) and count went down dramatically
- I've also ran `collectgarbage("collect")` couple of times to see if it makes any difference. It does reduce object count dramatically but only for very short period.

To me it looks like it's triggered by connecting external screen. Without any external screens connected ever, the count is pretty stable and memory usage is stable as well. I'm using https://github.com/awesomeWM/awesome/issues/1382#issuecomment-289378695 script to remap the windows to external screen (and back again).

Output of `awesome --version`:
```
awesome v4.3 (Too long)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2
```

I looked for usual suspects like `awful.spawn.with_line_callback` and I did have one instance of that (that forks acpi_listen) and removed it without any positive effects. I do use `awful.spawn.with_shell` and `awful.spawn` for the usual menus and hotkeys but they are not external screen specific and are constantly used.

I played around with step GC without any positive results.

Another debug output originally suggested:
```
$ awesome-client 'return require("gears.debug").dump_return{ button = button.instances(), client=client.instances(), drawable=drawable.instances(), drawin=drawin.instances(), key=key.instances(), screen=screen.instances(), tag=tag.instances() }'
   string "table: 0x5648ab9f7600
  key : 104 (number)
  tag : 130 (number)
  drawin : 14 (number)
  button : 2048 (number)
  screen : 13 (number)
  client : 155 (number)
  drawable : 14 (number)"
```

I've been looking for other things that might make a difference when external screens are connected and have found none. I run custom power/screen management scripts so there are no other daemons/processes that might silently do something behind the scenes.

Only other thing I've found that does not correlate to external screens in any way, this is the worst example:

`2020-01-10 10:19:40 W: awesome: a_glib_poll:438: Last main loop iteration took 1.572473 seconds! Increasing limit for this warning to that value.`

Any hints how to debug this are very much appreciated.

**Comment by psychon** [comment]
Could you provide a link to your config? I guess "does this also happen with the default config" is hard to test since this takes so long to manifest.

> Another debug output originally suggested:

Uhm. There are 13 screen instances, 130 tags (number of screens times 10, I guess). And 155 clients. All of these numbers seem too high.

Does this "13 screen instances" sound like "no screen was ever GC'd"?

This still does not really explain the way-too-high memory usage, but it sounds like a starting point.

Once upon a time, I hacked together the following script: https://gist.github.com/psychon/5a9c3dbd04a3a4d2589ec6f264696c10
Usage would be `require("traverse").new():do_all()`. What this does is to start at `_G` and `debug.getregistry()` and follow all reachable "stuff" that it finds. Everything that was seen is recorded in a table so that loops in data structures do not cause endless loops, but I cannot guarantee that this does not cause endless loops.

The version that I pasted above has `traverse.traverse_screen = traverse.traverse_unknown`. This causes it to call `traverse_unknown` when it comes across a screen. This function prints the "path" that was used to reach the screen object.

Long story short: This should allow figuring out why screens are still reachable and thus why they are not GC'd.

Can you try this out and see if this prints anything useful? The scenario should be something like:
- Connect and disconnect another screen
- run `awesome-client 'for i=1, 10 do collectgarbage("collect") end'
- (perhaps repeat the above two steps three times, just to be sure? dunno)
- Run something like `awesome-client 'dofile("/tmp/traverse.lua").new():do_all()'`
- Look at Awesome's stdout output and hope that it is helpful.

Does this sound like a plan?

**Comment by tarko** [comment]
I managed to kernelpanic my machine last night while hacking something so I have nice and clean setup to debug this on :)

> Could you provide a link to your config?

https://gist.github.com/tarko/203cb579fc14c4a97558672fbdd6625c

> Uhm. There are 13 screen instances, 130 tags (number of screens times 10, I guess). And 155 clients. All of these numbers seem too high.
> Does this "13 screen instances" sound like "no screen was ever GC'd"?

I should have caught that myself. This is indeed what seems to be happening, there should be only single screen and 10 tags.

> Can you try this out and see if this prints anything useful? The scenario should be something like:
> Does this sound like a plan?

It sure does.

With no external screens connected ever (I also added instance counts for comparison):

```
unknown screen  screen: 0x5584023a9128  _G[package][loaded][naughty.core][notifications].k(screen: 0x5584023a9128)
unknown thread  thread: 0x5584023bd5f8  registry[3]
unknown thread  thread: 0x5584023ee728  registry[userdata: 0x7f699b4a712c][2]

  button : 344 (number)
  client : 30 (number)
  key : 104 (number)
  tag : 10 (number)
  screen : 1 (number)
  drawable : 2 (number)
  drawin : 2 (number)"
```

After external screen has been connected few times and client/tag relocation forced by disabling eDP-1:

```
unknown screen  screen: 0x5584040fb588  _G[package][loaded][naughty.core][notifications].k(screen: 0x5584040fb588)
unknown screen  tostring-error  _G[mymainmenu][wibox][_drawable][_widget_context][screen]
unknown screen  tostring-error  _G[package][loaded][awful.tag][history][update]f:/usr/share/awesome/lib/awful/tag.lua:434(up:data)[history].k(tostring-error)
unknown screen  tostring-error  _G[package][loaded][awful.tag][history][update]f:/usr/share/awesome/lib/awful/tag.lua:434(up:data)[history].k(tostring-error)
unknown screen  tostring-error  _G[package][loaded][awful.tag][history][update]f:/usr/share/awesome/lib/awful/tag.lua:434(up:data)[history].k(tostring-error)
unknown thread  thread: 0x5584023bd5f8  registry[3]
unknown thread  thread: 0x5584023ee728  registry[userdata: 0x7f699b4a712c][2]


  button : 820 (number)
  client : 33 (number)
  key : 104 (number)
  tag : 50 (number)
  screen : 6 (number)
  drawable : 6 (number)
  drawin : 6 (number)"
```

First idea that comes to mind reading this is that tag history keeps all the old screen references alive.

**Comment by psychon** [comment]
> First idea that comes to mind reading this is that tag history keeps all the old screen references alive.

Yup. However, this only indicates that *at least* the tag history keeps old screens alive. This can easily turn into a game of whack-a-mole, where one fix only uncovers the next problem.

Anyway, the following causes an error (actually two errors, hence the new call to `gears.protected_call`):
```diff
diff --git a/tests/test-leaks.lua b/tests/test-leaks.lua
index fa185575b..66c368a92 100644
--- a/tests/test-leaks.lua
+++ b/tests/test-leaks.lua
@@ -20,6 +20,10 @@ for s in screen do
 end
 run_delayed_calls()
 
+local function safe_tostring(obj)
+    return require("gears.protected_call")(tostring, obj) or "error-during-tostring"
+end
+
 -- Test if some objects can be garbage collected
 local function collectable(a, b, c, d, e, f, g, h, last)
     assert(last == nil, "got more arguments than supported")
@@ -35,7 +39,7 @@ local function collectable(a, b, c, d, e, f, g, h, last)
     -- Check if the table is now empty
     for _, v in pairs(objs) do
         print("Some object was not garbage collected!")
-        error(v)
+        error(safe_tostring(v))
     end
 end
 
@@ -81,6 +85,18 @@ collectable(awful.widget.tasklist{screen=1, filter=awful.widget.tasklist.filter.
 prepare_for_collect = run_delayed_calls
 collectable(create_wibox())
 
+-- Test that screens can be collected
+local function create_and_remove_screen()
+    local s = screen.fake_add(-10, -10, 10, 10)
+    run_delayed_calls()
+    awful.tag.history.update(s)
+    run_delayed_calls()
+    s:fake_remove()
+    return s
+end
+prepare_for_collect = run_delayed_calls
+collectable(create_and_remove_screen())
+
 runner.run_steps({ function() return true end })
 
 -- vim: filetype=lua:expandtab:shiftwidth=4:tabstop=8:softtabstop=4:textwidth=80
```
A possible fix is almost trivial:
```diff
diff --git a/lib/awful/tag.lua b/lib/awful/tag.lua
index 8874d69c7..28ef00c03 100644
--- a/lib/awful/tag.lua
+++ b/lib/awful/tag.lua
@@ -1676,6 +1676,7 @@ capi.screen.connect_signal("removed", function(s)
             t.data.awful_tag_properties.screen = nil
         end
     end
+    data.history[s] = nil
 end)
 
 function tag.mt:__call(...)
```
I do not have much time currently. Anyone wants to come up with a slightly less ugly test and submit a PR with a test and a fix? (Also: Are these calls to `run_delayed_calls()` really necessary? I would expect not and a quick test confirms that "yes").

Actually... a better test would be to call `awful.tag.viewnext(s)` instead of `awful.tag.history.update(s)`. This still reproduces the original problem and "tests more".

**Comment by psychon** [comment]
Oh, I forgot to write: @tarko can you test the above trivial patch to `awful.tag` and see if it makes "stuff work"?

**Comment by psychon** [comment]
(Sorry for triple-posting)

> This can easily turn into a game of whack-a-mole, where one fix only uncovers the next problem.

Actually... the tag history does *not* explain why the client objects cannot be collected. And we actually do have a dedicated test to ensure that clients are GC'able: `test-leak-client.lua`. So apparently that test does not do something that is necessary to cause the leak (just as we so far have no tests for the tag history with leaks).
(I tested adding some calls to `awful.tag.viewnext()` in `test-leak-client.lua` and this did not cause the tests to fail)

**Comment by tarko** [comment]
Whack-a-mole indeed. After forcing GC I'm left with:

```
unknown screen  screen: 0x55dc82362768  _G[package][loaded][naughty][notifications].k(screen: 0x55dc82362768)
unknown screen  tostring-error  _G[mymainmenu][wibox][_drawable][_widget_context][screen]
unknown thread  thread: 0x55dc7f8b8c78  registry[3]
unknown thread  thread: 0x55dc7f8e8048  registry[userdata: 0x7f6fa56c112c][2]

  drawin : 3 (number)
  screen : 2 (number)
  client : 11 (number)
  drawable : 3 (number)
  key : 104 (number)
  button : 476 (number)
  tag : 20 (number)"

```

**Comment by psychon** [comment]
>  _G[mymainmenu][wibox][_drawable][_widget_context][screen]

Well... I'm not sure if this counts as a leak. You `mymainmenu` was last visible on the screen that was removed. The next redraw would update it for another screen. I do not think adding extra magic to "speed this up" is a good idea.
(also, there can only be a single "last visible on screen", so this can only keep a single screen object alive - but still, this could be bad if you had many different menus and only used them rarely...)

**Comment by tarko** [comment]
Sure, as long as I'm left with screens*2 and tags*2 forever then this should keep memory usage under control. I'll keep monitoring and let's see what happens.

**Comment by psychon** [comment]
I just opened #2991 for the tag history screen leak.

@tarko By the way, you also might want to monitor the number client instances and see if that stays suspiciously high. If you want to debug that, you can change `traverse.traverse_client = nop` into `traverse.traverse_client = traverse.traverse_unknown` in my `traverse.lua` (even though this might be quite spammy for many clients where some of them are not even closed...)

**Comment by tarko** [comment]
@psychon Ack, I was on it already and found that my cyclefocus setup is also leaking clients, yet to investigate is this due to my modifications or out-of-the-box.

**Comment by tarko** [comment]
Following this one up with another observation. It looks like client maximize/minimize etc. attributes also leak screens (I don't use them very much so leak is much slower in my case):

```
unknown screen  tostring-error  _G[package][loaded][awful.placement](mt)[__index][restore](mt)[__call]f:/usr/share/awesome/lib/awful/placement.lua:220(up:f)f:/usr/share/awesome/lib/awful/placement.lua:1797(up:data)[tostring-error][maximize_vertically][screen]
unknown screen  tostring-error  _G[package][loaded][awful.placement](mt)[__index][restore](mt)[__call]f:/usr/share/awesome/lib/awful/placement.lua:220(up:f)f:/usr/share/awesome/lib/awful/placement.lua:1797(up:data)[tostring-error][maximize_vertically][screen]
unknown screen  tostring-error  _G[package][loaded][awful.placement](mt)[__index][restore](mt)[__call]f:/usr/share/awesome/lib/awful/placement.lua:220(up:f)f:/usr/share/awesome/lib/awful/placement.lua:1797(up:data)[tostring-error][maximize][screen]
unknown screen  tostring-error  _G[package][loaded][awful.placement](mt)[__index][restore](mt)[__call]f:/usr/share/awesome/lib/awful/placement.lua:220(up:f)f:/usr/share/awesome/lib/awful/placement.lua:1797(up:data)[tostring-error][maximize_vertically][screen]
unknown screen  tostring-error  _G[package][loaded][awful.placement](mt)[__index][restore](mt)[__call]f:/usr/share/awesome/lib/awful/placement.lua:220(up:f)f:/usr/share/awesome/lib/awful/placement.lua:1797(up:data)[tostring-error][maximize][screen]
...
```

**Comment by psychon** [comment]
Okay, since this is about `awful.placement`, let's digest it:
- `[awful.placement](mt)[__index]` is the `placement_private` table in `awful.placement`.
- `[restore](mt)[__call]` is the extra wrapper that `wrap_client` puts around a function
- No idea which function `f:/usr/share/awesome/lib/awful/placement.lua:220(up:f)` is since line numbers became useless
- Same for `f:/usr/share/awesome/lib/awful/placement.lua:1797`
- `(up:data)` would be this, I guess (which is only accessed by `store_geometry` and `restore`... I would guess the second function above was `restore` and the first one is the anonymous function that `wrap_client` creates to discard the first argument): https://github.com/awesomeWM/awesome/blob/2da1cb9ba04eb223bd0412f64ad80a8103c9aa0b/lib/awful/placement.lua#L287
- `[tostring-error]` this must be an object (a client?) that was already destroyed/unmanaged
- The entry for this client then keeps the screen alive

Uhm... the `data` table should have weak keys, so in theory the GC should be able to clean this up. Did you run `collectgarbage("collect")` (more than once) before getting this output? If no, then I would guess that this is only because the GC did not manage to collect these clients yet. If yes, then there might be a reference cycle here. I think weak tables turn into memory leaks if the value references the key (here: If the screen would also keep a reference to the dead client, which could very well be the case).

**Comment by tarko** [comment]
Correct, I ofc ran GC (10 times) before taking this snapshot.

**Comment by psychon** [comment]
@Elv13 What do you think about adding a `client.connect_signal("unmanage", ` for clearing the "placement cache"?
